### ISSUE TITLE
Bad request crash while connecting

### ISSUE THREAD

**Issue by Thun0** [librespot-org/librespot#257]
Hey,
I tried to use librespot on Raspberry Pi 3 and I keep getting error 400 for my user URI. This happens when I pick my device to connect to in Spotify client.

```Oct 29 21:57:59 superbia librespot[978]: INFO:librespot_core::session: Connecting to AP "gew1-accesspoint-b-v0r4.ap.spotify.com:4070"
Oct 29 21:57:59 superbia librespot[978]: INFO:librespot_core::session: Authenticated as "<username>" !
Oct 29 21:57:59 superbia librespot[978]: INFO:librespot_playback::audio_backend::alsa: Using alsa sink
Oct 29 21:57:59 superbia librespot[978]: INFO:librespot_core::session: Country: "PL"
Oct 29 21:58:00 superbia librespot[978]: WARN:librespot_core::mercury: error 400 for uri hm://remote/3/user/<username>/
Oct 29 21:58:00 superbia librespot[978]: WARN:librespot_core::mercury: error 400 for uri hm://remote/3/user/<username>/
Oct 29 21:58:00 superbia librespot[978]: thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: MercuryError', libcore/result.rs:945:5
Oct 29 21:58:00 superbia librespot[978]: stack backtrace:
Oct 29 21:58:00 superbia librespot[978]:    0:   0x99872b - std::sys::unix::backtrace::tracing::imp::unwind_backtrace::hcf3dddc6200fad89
Oct 29 21:58:00 superbia librespot[978]:                        at libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
Oct 29 21:58:00 superbia librespot[978]:    1:   0x986e3f - std::sys_common::backtrace::print::ha74183f8146bff60
Oct 29 21:58:00 superbia librespot[978]:                        at libstd/sys_common/backtrace.rs:71
Oct 29 21:58:00 superbia librespot[978]:                        at libstd/sys_common/backtrace.rs:59
Oct 29 21:58:00 superbia librespot[978]:    2:   0x984957 - std::panicking::default_hook::{{closure}}::h8790cb9f07a765c6
Oct 29 21:58:00 superbia librespot[978]:                        at libstd/panicking.rs:211
Oct 29 21:58:00 superbia librespot[978]:    3:   0x9845d7 - std::panicking::default_hook::h8932bf45ef21858c
Oct 29 21:58:00 superbia librespot[978]:                        at libstd/panicking.rs:227
Oct 29 21:58:00 superbia librespot[978]:    4:   0x98507f - std::panicking::rust_panic_with_hook::hc91e5f2c01fc689c
Oct 29 21:58:00 superbia librespot[978]:                        at libstd/panicking.rs:475
Oct 29 21:58:00 superbia librespot[978]:    5:   0x984c43 - std::panicking::continue_panic_fmt::h4a8945dc115e38fc
Oct 29 21:58:00 superbia librespot[978]:                        at libstd/panicking.rs:390
Oct 29 21:58:00 superbia librespot[978]:    6:   0x984b2b - rust_begin_unwind
Oct 29 21:58:00 superbia librespot[978]:                        at libstd/panicking.rs:325
Oct 29 21:58:00 superbia librespot[978]:    7:   0x9e902f - core::panicking::panic_fmt::h26b7cf0675909ac6
Oct 29 21:58:00 superbia librespot[978]:                        at libcore/panicking.rs:77
Oct 29 21:58:00 superbia librespot[978]:    8:   0x549b33 - core::result::unwrap_failed::h9c019f0afa067342
Oct 29 21:58:00 superbia librespot[978]:    9:   0x544153 - <librespot_connect::spirc::SpircTask as futures::future::Future>::poll::hf2bbf60e5e0fcfc2
Oct 29 21:58:00 superbia librespot[978]:   10:   0x524433 - <librespot::Main as futures::future::Future>::poll::h3220540858f77eb6
Oct 29 21:58:00 superbia librespot[978]:   11:   0x50482f - futures::task_impl::std::set::hba6f3426ce6056a3
Oct 29 21:58:00 superbia librespot[978]:   12:   0x51e83f - <futures::future::lazy::Lazy<F, R> as futures::future::Future>::poll::h39a50ce9fc49b659
Oct 29 21:58:00 superbia librespot[978]:   13:   0x50451f - futures::task_impl::std::set::h3df2553c79eb4dc5
Oct 29 21:58:00 superbia librespot[978]:   14:   0x50b6a3 - <std::thread::local::LocalKey<T>>::with::h39e96f1b9679251f
Oct 29 21:58:00 superbia librespot[978]:   15:   0x50f7c3 - <tokio::executor::current_thread::Entered<'a, P>>::block_on::h91dba97a87e364e7
Oct 29 21:58:00 superbia librespot[978]:   16:   0x50ba33 - <std::thread::local::LocalKey<T>>::with::haf1881f2b5f00576
Oct 29 21:58:00 superbia librespot[978]:   17:   0x50b4d7 - tokio_reactor::with_default::h2ae530e1c5a805bc
Oct 29 21:58:00 superbia librespot[978]:   18:   0x50e6c3 - <scoped_tls::ScopedKey<T>>::set::he35c17dbd7d05fb6
Oct 29 21:58:00 superbia librespot[978]:   19:   0x506d2f - tokio_core::reactor::Core::run::h13d589b7351a75df
Oct 29 21:58:00 superbia librespot[978]:   20:   0x52641b - librespot::main::h7263f0722bde5bef
Oct 29 21:58:00 superbia librespot[978]:   21:   0x50b597 - std::rt::lang_start::{{closure}}::hda8c349478babf2e
Oct 29 21:58:00 superbia librespot[978]:   22:   0x984aa3 - std::panicking::try::do_call::hc221fc47d38c79d7
Oct 29 21:58:00 superbia librespot[978]:                        at libstd/rt.rs:59
Oct 29 21:58:00 superbia librespot[978]:                        at libstd/panicking.rs:310
Oct 29 21:58:00 superbia librespot[978]:   23:   0x9a4c8f - __rust_maybe_catch_panic
Oct 29 21:58:00 superbia librespot[978]:                        at libpanic_unwind/lib.rs:105
Oct 29 21:58:00 superbia librespot[978]:   24:   0x981d5b - std::rt::lang_start_internal::h8c7a5dbe81fee471
Oct 29 21:58:00 superbia librespot[978]:                        at libstd/panicking.rs:289
Oct 29 21:58:00 superbia librespot[978]:                        at libstd/panic.rs:392
Oct 29 21:58:00 superbia librespot[978]:                        at libstd/rt.rs:58
Oct 29 21:58:00 superbia librespot[978]:   25:   0x527083 - main
Oct 29 21:58:00 superbia librespot[978]:   26: 0x76cab677 - __libc_start_main
```

I don't know if it matters but my username contains '|' (pipe) character.

**Comment by Thun0** [comment]
If I URLencode my username, then librespot gives authentication error :/

**Comment by ashthespy** [comment]
I have a snippet somewhere floating around - but this needs to be done at: https://github.com/librespot-org/librespot/blob/431be9e847a584d374ec44e7c138525dd8fbb9d6/connect/src/spirc.rs#L217

**Comment by Thun0** [comment]
Thanks! It passed the request part. Now I have problem with backend. No matter which backend I choose during build. I cannot use anything else than "pipe" or "portaudio" when running. It gives me error on invalid backend :/

**Comment by ashthespy** [comment]
Shouldn't it just be `alsa`?
Check out the [wiki](https://github.com/librespot-org/librespot/wiki/Options) for more details on the options

**Comment by Thun0** [comment]
Yes, it is. It was a snippet from my random tries. It still doesn't work. Picked "alsa-backend" in cargo build and --backend alsa during run.

**Comment by ashthespy** [comment]
Hmm, what do mean by doesn't work? This patch should not effect the backend/sink system in anyway.

**Comment by roderickvd** [comment]
Fixed by #476.
